### PR TITLE
Align KTO with DPO: Support PEFT with Liger

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Support PEFT with Liger.

Follow-up to:
- #6159 

Part of:
- #4786

This PR enables training PEFT models with the Liger fused KTO loss in the experimental KTOTrainer, aligning it with the DPO trainer (#6159).

### Motivation

Previously the experimental KTOTrainer rejected any combination of PEFT and use_liger_kernel=True with a blanket error, even though the common case (a weight-based adapter such as LoRA that does not target lm_head) works correctly. DPO already supports this combination, so KTO was needlessly more restrictive.

### Solution

The blanket restriction is replaced with real support in the Liger loss path. When the model is a PEFT model, the trainer unwraps it to the underlying backbone before invoking the fused loss, and recovers reference-model behaviour from the frozen adapter when no explicit reference model is provided (matching the logit-based reference computation).

Two configurations remain genuinely incompatible with the fused loss and now fail loudly instead of being blanket-rejected or, worse, silently mistraining:

- An adapter targeting lm_head: the Liger loss reads lm_head.weight directly, so the head adapter would receive no gradient and train as if frozen.
- Prompt-learning methods (PromptTuning, PrefixTuning, P-Tuning): they inject virtual tokens via PeftModel.forward(), which the fused loss bypasses, so the loss would be computed on the wrong sequence.

### Changes

- Remove the blanket error forbidding PEFT models when use_liger_kernel=True.
- Add PEFT support to the Liger KTO loss path, including reference-model recovery via the frozen adapter when no reference model is supplied.
- Raise a clear error when a PEFT adapter targets lm_head under the Liger loss.
- Raise a clear error when a prompt-learning PEFT method is combined with the Liger loss.
- Replace the old init-fails test with a test that verifies the full PEFT + Liger training path (adapter params update, base params stay frozen).
- Add tests covering the lm_head-adapter and prompt-learning error cases.

---

> [!NOTE]
> **Medium Risk**
> Changes core training loss wiring for PEFT+Liger (reference adapter switching and backbone unwrapping); wrong behavior could silently affect alignment runs, though incompatible configs now fail fast.
> 
> **Overview**
> **Experimental `KTOTrainer` now allows `use_liger_kernel=True` with weight-based PEFT (e.g. LoRA that does not target `lm_head`)**, matching DPO behavior instead of rejecting all PEFT+Liger combinations.
> 
> Init-time checks are narrowed: **`lm_head` adapters** and **prompt-learning PEFT** (PromptTuning, etc.) still raise clear `ValueError`s because the fused loss reads `lm_head.weight` directly and bypasses `PeftModel.forward()`.
> 
> The **Liger loss path** unwraps PEFT models to the inner backbone for the forward pass, and when no separate `ref_model` is set it **derives reference hidden states via the frozen `ref` adapter** (same idea as the logit-based path). Tests add an end-to-end LoRA+Liger train case and separate tests for the two incompatible PEFT setups.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bfa6eb16c8a97ac71a5ce89abf2bc39cf41b87ab. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by qgallouedec — APPROVED** [review]
exactly the same as https://github.com/huggingface/trl/pull/6159, thanks

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -1255,26 +1271,55 @@ def _compute_loss_liger(self, model, inputs, return_outputs):
...
+                else:
+                    ref_backbone = ref_model_inner.base_model
+                ref_outputs = ref_backbone(**model_kwargs)
+                ref_lm_head = self.ref_model.get_output_embeddings()
+
+        if self.calculate_KL:
+            kl = (KL_logps - ref_KL_logps).mean().detach()
+            kl = self.accelerator.gather_for_metrics(kl).mean().clamp(min=0)
+        else:
+            kl = torch.zeros(1).to(self.accelerator.device)
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:1322]: Addressed in PR:
- #6279
